storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +249,187 @@ func (s *commitLogSource) Read(
-	blocksPool := bopts.DatabaseBlockOptions().DatabaseBlockPool()
-	multiReaderIteratorPool := blopts.MultiReaderIteratorPool()
 	for shard, unmergedShard := range unmerged {
-		shardResult := result.NewShardResult(len(unmergedShard), s.opts.ResultOptions())
-		for _, unmergedBlocks := range unmergedShard {
-			blocks := block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
-			for start, unmergedBlock := range unmergedBlocks.encoders {
-				block := blocksPool.Get()
-				if len(unmergedBlock) == 0 {
-					emptyErrs++
-					continue
-				} else if len(unmergedBlock) == 1 {
-					block.Reset(start, unmergedBlock[0].enc.Discard())
-				} else {
-					readers := make([]io.Reader, len(unmergedBlock))
-					for i := range unmergedBlock {
-						stream := unmergedBlock[i].enc.Stream()
-						readers[i] = stream
-						defer stream.Finalize()
-					}
+		mergeSemaphore <- true
+		wg.Add(1)
+		go func(shard int, unmergedShard encodersAndRanges) {
+			var shardResult result.ShardResult
+			shardResult, shardEmptyErrs[shard], shardErrs[shard] = s.mergeShard(
+				unmergedShard, blocksPool, multiReaderIteratorPool, encoderPool, blopts)
+			if len(shardResult.AllSeries()) > 0 {
+				// Prevent race conditions while updating bootstrapResult from multiple go-routines
+				bootstrapResultLock.Lock()
+				// Shard is a slice index so conversion to uint32 is safe
+				bootstrapResult.Add(uint32(shard), shardResult, nil)
+				bootstrapResultLock.Unlock()
+			}
+			<-mergeSemaphore
+			wg.Done()
+		}(shard, unmergedShard)
+	}
 
-					iter := multiReaderIteratorPool.Get()
-					iter.Reset(readers)
-
-					var err error
-					enc := encoderPool.Get()
-					enc.Reset(start, blopts.DatabaseBlockAllocSize())
-					for iter.Next() {
-						dp, unit, annotation := iter.Current()
-						encodeErr := enc.Encode(dp, unit, annotation)
-						if encodeErr != nil {
-							err = encodeErr
-							errs++
-							break
-						}
-					}
-					if iterErr := iter.Err(); iterErr != nil {
-						if err == nil {
-							err = iter.Err()
-						}
-						errs++
-					}
+	// Wait for all merge goroutines to complete
+	wg.Wait()
+	s.logMergeShardsOutcome(shardErrs, shardEmptyErrs)
+	return bootstrapResult
+}
 
-					iter.Close()
-					for i := range unmergedBlock {
-						unmergedBlock[i].enc.Close()
-					}
+func (s *commitLogSource) mergeShard(
+	unmergedShard encodersAndRanges,
+	blocksPool block.DatabaseBlockPool,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	encoderPool encoding.EncoderPool,
+	blopts block.Options,
+) (shardResult result.ShardResult, numShardEmptyErrs int, numErrs int) {
+	shardResult = result.NewShardResult(len(unmergedShard.encodersBySeries), s.opts.ResultOptions())
+	for _, unmergedBlocks := range unmergedShard.encodersBySeries {
+		blocks := block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
+		for start, unmergedBlockEncoders := range unmergedBlocks.encoders {
+			startInNano := time.Unix(0, start)
+			block := blocksPool.Get()
```

Comment:
there's a couple of continues in the loops below which skip over the block, should we be putting the block back into the pool in those cases?

Reply:
I think you're right. The block is only added to "blocks" if no error occurs, so if any error occurs and we continue, then it should be safe to return it to the pool. I'll add that.

Reply:
Also it looks like the iterator is never returned to the multiiteratorpool after its done. Its safe to do so once the Close() method is called right?

Reply:
Nevermind the Close() method will return it to the pool